internal/docstore/mongodocstore/mongo.go

```diff
@@ -209,6 +259,29 @@ func (c *collection) update(ctx context.Context, a *driver.Action) error {
 	return nil
 }
 
+func (c *collection) doUpdate(ctx context.Context, filter bson.D, updateDoc map[string]bson.D, upsert bool) (*mongo.UpdateResult, error) {
+	opts := options.Update()
+	if upsert {
+		opts.SetUpsert(true) // Document will be created if it doesn't exist.
+	}
+	// Always add or increment the revision field of the stored document, even if the argument doc
+	// didn't have one.
+	updateDoc["$inc"] = bson.D{{Key: docstore.RevisionField, Value: 1}}
+	return c.coll.UpdateOne(ctx, filter, updateDoc, opts)
+}
+
+// Error code for a write error when no documents match a filter.
+// (The Go mongo driver doesn't define an exported constant for this.)
+const mongoNotFoundCode = 11000
+
 func (c *collection) ErrorCode(err error) gcerrors.ErrorCode {
-	return gcerr.GRPCCode(err)
+	if c := gcerr.GRPCCode(err); c != gcerrors.Unknown {
+		return c
+	}
+	if wexc, ok := err.(mongo.WriteException); ok && len(wexc.WriteErrors) > 0 {
```

Comment:
It looks like all `err`s have `gcerrors.Unknown` codes?

Reply:
Not sure exactly what you mean, but I shouldn't be using GRPCCode here. Fixed.